core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,59 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      inLock(stateLock) {
```

Comment:
Thinking about it a bit more, we do not need to whole the lock when appending txn messages to local logs as we discussed in another PR: https://github.com/apache/kafka/pull/3082#discussion_r117588659

Plus, holding the lock here may result in a dead lock situation, since in purgatory we hold the delayed producer object when calling `tryComplete` which will call the `removeFromCacheCallback` below. So there can be thread 1 trying to grab `statelock` first then the delayed producer `object` lock later in `tryCompleteElseWatch`, while thread 2 trying to grab the `object` lock first in `tryComplete` when an fetch request is handled and hence HW has been advanced, and then inside `removeFromCacheCallback` trying to grab the `statelock`, hence the deadlock.

We can consider 1) only grab the lock while getting the list of txnIds to expire along with its belonged partitions, 2) release the lock and then try to append to log.